framework/src/application/node/forger/forger.js

```diff
@@ -220,7 +233,56 @@ class Forger {
 					)} is not a delegate`,
 				);
 			}
+			// Prepare hash-onion
+			const registeredHashOnion = registeredHashOnions[account.address];
+			const { hashOnion: hashOnionConfig } =
+				this.config.forging.delegates.find(
+					d => getAddressFromPublicKey(d.publicKey) === account.address,
+				) || {};
+			if (!hashOnionConfig) {
+				throw new Error(
+					`Account ${account.address} does not have hash onion in the config`,
+				);
+			}
+			const configHashOnionSeed =
+				hashOnionConfig.hashes[hashOnionConfig.hashes.length - 1];
+			if (registeredHashOnion && registeredHashOnion !== configHashOnionSeed) {
+				this.logger.warn(
+					`Hash onion for Account ${account.address} is not the same as previous one. Overwriting with new hash onion`,
+				);
+				usedHashOnions = usedHashOnions.filter(
+					ho => ho.address !== account.address,
+				);
+			}
+			registeredHashOnions[account.address] = configHashOnionSeed;
+			const highestUsedHashOnion = usedHashOnions.reduce((prev, current) => {
+				if (current.address !== account.address) {
+					return prev;
+				}
+				if (!prev || prev.count < current.count) {
+					return current;
+				}
+				return prev;
+			}, undefined);
```

Comment:
This part is pretty confusing. Can you explain the data structure for `usedHashOnions`? 

Why not use simply save the onion seed counter for every delegate as we already storing just the seed. This way we can just pick up the next hash from respective seed onion.

Reply:
usedHashOnion is just an array of object `{ height: number, address: string, count: number }`.
We cannot simply keep the counter because when the block is deleted, we need to revert the counter, but forger info is out of block context. Therefore, it needs to keep what it used, and clean up when finality kicks in